ens-improvement-proposals/ensip-17-evmchain-reverse-resolution.md

```diff
@@ -0,0 +1,92 @@
+
+
+## Specification
+
+### Overview
+
+* Reverse registrars will be setup on each L2, with a corresponding registry
+* Reverse registrar will only allow setting the name, without resolver customisability. This is to allow a consistent storage slot that can be checked on L1.
+* User can now claim their reverse on this L2 and set it to their ENS name
+* Their ENS name will need to set their record for the same EVM-cointype as the network, which is specified in [ENSIP-11](https://docs.ens.domains/ensip/11).
+* A dapp can then detect the chainID that a user is on, find the corresponding cointype and resolve their primary ens name by resolving the name record at [userAddress].[coinType].reverse. This will be resolved via CCIP-read and look up the reverse record on the corresponding EVM-chain.
+* Dapp will then resolve this name via ENS on L1 to check if the forward resolution matches. This forward resolution can be on L1, or the user can set up CCIP-read records for each network and put those addresses wherever they want.
+* Once matched, the dapp can then also resolve any text records on the primary ENS name, such as avatar.
+
+### Resolving Primary ENS Names by a dapp
+
+1) If a dapp has a connected user it SHOULD first check the chainId of the user.
+2) If the chainId is not 1, it SHOULD then construct the ENS name [address].[coinType].reverse to obtain the primary ENS name of the user. 
+3) If none is found, it SHOULD check check mainnet [address].default.reverse.
+4) If the dapp finds an ENS name, it MUST first check the forward resolution. The forward resolution MUST match the same coin type as the chain id of the user.
+
+Note: The dapp MUST NOT use the reverse record set for mainnet even if the Primary ENS name has not been set on the target chain, and must instead show the address.
+
+### Resolving an avatar by a dapp on L2
+
+ENSIP-12 was concieved before the ENS L2 reverse resolution specification and therefore should be updated to reflect the current state of ENS primary name resolution. This means that all avatar records are able to be updated on a per-chain basis by updating the avatar record on their reverse node.
+
+#### Example for an EVM Address
+
+To determine the avatar URI for a specific EVM chain address, the client MUST reverse-resolve the address by querying the ENS registry on Ethereum for the resolver of `<address>.<coinType>.reverse`, where `<address>` is the lowercase hex-encoded Ethereum address, without leading '0x'. Then, the client calls `.text(namehash('<address>.<coinType>.reverse'), 'avatar')` to retrieve the avatar URI for the address.
+
+If a resolver is returned for the reverse record, but calling `text` causes a revert or returns an empty string, the client MUST call `.name(namehash('<address>.<coinType>.reverse'))`. If this method returns a valid ENS name, the client MUST:
```

Comment:
The way this is worded is overly specific, and implies a method that won't work for a default reverse resolver.

Reply:
Yeah actually this doesn't really make sense, because there are no custom resolvers on L2 reverse registrars anyway

Reply:
Also, the examples should be written descriptively - describing what the client does in the example scenario. SHOULd and MUST etc are for the specification itself.